shared/timeutils/timeutils.h

```diff
@@ -27,6 +27,46 @@
 #ifndef MICROPY_INCLUDED_LIB_TIMEUTILS_TIMEUTILS_H
 #define MICROPY_INCLUDED_LIB_TIMEUTILS_TIMEUTILS_H
 
+#if MICROPY_PY_BUILTINS_FLOAT && MICROPY_FLOAT_IMPL == MICROPY_FLOAT_IMPL_DOUBLE
+#include <math.h> // required for trunc()
+#endif
+#include "py/obj.h"
+
+// To maintain reasonable compatibility with CPython on embedded systems,
+// and avoid breaking anytime soon, time functions are defined to work
+// at least between 1970 and 2099 on any machine.
+//
+// Specific ports can set MP_SUPPORT_Y2100_AND_BEYOND to 1 in order
+// to get extended date support beyond 2099
+//
+// By default this is enabled for machines using 64 bit pointers only,
+// but it can be enabled by specific ports
+#ifndef MP_SUPPORT_Y2100_AND_BEYOND
+#if MP_SSIZE_MAX > 2147483647
+#define MP_SUPPORT_Y2100_AND_BEYOND 1
+#else
+#define MP_SUPPORT_Y2100_AND_BEYOND 0
+#endif
+#endif
+
+#define MICROPY_EPOCH_IS_2000 (!(MICROPY_EPOCH_IS_1970))
+
+// `platform_timestamp_t` is the type that should be used by the port
+// to represent timestamps, and is referenced to the platform epoch
+#if MP_SUPPORT_Y2100_AND_BEYOND || MICROPY_EPOCH_IS_2000
+typedef int64_t platform_timestamp_t;
+#else
+typedef mp_uint_t platform_timestamp_t;
```

Comment:
This type is used in `extmod` code so (due to code layering) I think it's better to define it in the core in `py/mpconfig.h`.

And also call it `mp_platform_timestamp_t` or even just `mp_timestamp_t` (that matches eg `mp_off_t`).